### PR TITLE
Discord notifications for health-check integration

### PR DESCRIPTION
**Pull request by kwypchlo**
### Features

- integrated with Discord via discord.js library, uses DISCORD_TOKEN from environment variables
- sends a hello message on startup (just to know it ran)
- sends a message every time a critical or verbose checks changed state (so for example when we had 20/20 passing and now we have 18/20 passing but also if we go back to 20/20 passing) 
- it will skip unnecessary repetitions of the same message (for example when we have 20/20 passing and nothing is changing but also when we have 18/20 passing and nothing is changing) but will send a message 24 hours from the last message sent anyway just to tell us it's alive

### TODO

- comments

### CODE REVIEW

File: packages/health-check/src/notifyChanges.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+const db = require("./db");
+const { sendMessageToHealthCheckChannel, getRoleByName } = require("./discord");
+
+const { PORTAL_NAME = "PORTAL_NAME not defined" } = process.env;
+
+const notificationTimestamps = { critical: 0, verbose: 0 };
+const notificationUnchangedInterval = 24 * 60 * 60 * 1000; // 24 hours
+
+function notifyChanges(type) {
```
Comment (peterjanbrone) [inline packages/health-check/src/notifyChanges.js:9]: The code is definitely readable but some comments would be nice. A small introduction on what it is you're trying to achieve.

File: packages/health-check/src/notifyChanges.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+const db = require("./db");
+const { sendMessageToHealthCheckChannel, getRoleByName } = require("./discord");
+
+const { PORTAL_NAME = "PORTAL_NAME not defined" } = process.env;
+
+const notificationTimestamps = { critical: 0, verbose: 0 };
+const notificationUnchangedInterval = 24 * 60 * 60 * 1000; // 24 hours
```
Comment (peterjanbrone) [inline packages/health-check/src/notifyChanges.js:7]: I would make it 8.

File: packages/health-check/src/notifyChanges.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  const currentPassing = currentEntry.checks.filter(({ up }) => up).length;
+  const currentTotal = currentEntry.checks.length;
+  const sendMessage = () => {
+    notificationTimestamps[type] = Date.now();
+
+    const message = `${PORTAL_NAME}: ${currentPassing} / ${currentTotal} ${type} checks passing`;
+
+    if (currentPassing === currentTotal) return sendMessageToHealthCheckChannel(message);
+
+    const role = getRoleByName("skynet-prod");
```
Comment (peterjanbrone) [inline packages/health-check/src/notifyChanges.js:23]: I think we have to make all of the Skynet related things configurable. This script is not very useful to anyone besides us, unless they tweak it.

**Review by peterjanbrone — APPROVED** [review]
It feels a bit rushed. Don't get me wrong I think it's great and very useful, but we should try and keep up the code quality as well. This repo could use a linter for example, there's no documentation or explanation, and it contains references that are specific to our setup.

File: packages/health-check/src/discord.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+const { Client, MessageMentions } = require("discord.js");
+
+const client = new Client();
+
+const { DISCORD_BOT_TOKEN, PORTAL_NAME = "PORTAL_NAME not defined" } = process.env;
+
+client.on("ready", () => {
+  sendMessageToHealthCheckChannel(`${PORTAL_NAME}: reporting for duty!`);
```
Comment (ro-tex) [inline packages/health-check/src/discord.js:8]: This happens only once, on portal startup, right?

File: packages/health-check/src/discord.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    console.log(`Channel ${channelName} not found!`);
+
+    return;
+  }
+
+  channel.send(message);
+}
+
+function sendMessageToHealthCheckChannel(message) {
+  return sendMessage(message, "skynet-portal-health-check");
```
Comment (ro-tex) [inline packages/health-check/src/discord.js:37]: This should be in a config file as well. I'm with PJ about https://github.com/NebulousLabs/skynet-webportal/pull/396/files#r487000246 - all things should be configurable. Maybe with a config file, maybe we env vars.

File: packages/health-check/src/discord.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+function sendMessage(message, channelName) {
+  const channel = client.channels.cache.find((channel) => channel.name === channelName);
+
+  if (!channel) {
+    console.log(`Channel ${channelName} not found!`);
+
+    return;
+  }
+
+  channel.send(message);
```
Comment (ro-tex) [inline packages/health-check/src/discord.js:33]: Shouldn't this an async operations and therefore awaited?

File: packages/health-check/src/notifyChanges.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  if (!currentEntry) return;
+
+  const currentPassing = currentEntry.checks.filter(({ up }) => up).length;
+  const currentTotal = currentEntry.checks.length;
+  const sendMessage = () => {
+    notificationTimestamps[type] = Date.now();
+
+    const message = `${PORTAL_NAME}: ${currentPassing} / ${currentTotal} ${type} checks passing`;
+
+    if (currentPassing === currentTotal) return sendMessageToHealthCheckChannel(message);
```
Comment (ro-tex) [inline packages/health-check/src/notifyChanges.js:21]: We seem to always send the healthy status. Why not follow the same rules as "status unchanged"?

File: packages/health-check/src/discord.js (hunk 4 above)
Comment (kwypchlo) [reply packages/health-check/src/discord.js:8]: No, this happens when health-check service starts.

Comment (ro-tex) [reply packages/health-check/src/discord.js:8]: Right, my question was more along the lines of is this a permanently running service that we'd start once and have running for a long time or is it going to be a short-lived one that would start up, do the checks, and then exit until it's time for it to run again.

**Comment by kwypchlo** [comment]
Closing this since we are going to pursue different approach

**Pull request closed without merging**